### ISSUE TITLE
new Treewalker property to skip all nested Modules if there is parsing exception

### ISSUE THREAD

**Issue by romani** [checkstyle/checkstyle#12542]
It is our biggest limitation to not work on non compilable files
https://checkstyle.org/writingchecks.html#Limitations
> Code has to be compilable by javac to get valid violations. If it is not, you can get hard to understand parse errors. 

All of this is blocker for users:
https://github.com/checkstyle/checkstyle/issues?q=is%3Aopen+is%3Aissue+label%3Aantlr
but in most cases they mostlikely to ignore all Checkstyle activity on such files and let them use fancy new jdk syntax for business reasons.

We rely on plugin to cover this problem from us
But we still have problems with new syntax of Java

We need new property like `skipFileOnJavaParseException` or some thing similar.
If user wants new features of Java and ok to skip checkstyle in such files, it should be easy to do
We should not limit our users and let them decide : choose one checkstyle of new jdk features
By default it is disabled so exception crash all as before
But we can hint users after they create issue on use, about it, and they will be unblocked

We will introduce a new property called `skipFileOnJavaParseException`:
```
<module name="TreeWalker">
  <property name="skipFileOnJavaParseException" value="false"/>

```

**Comment by romani** [comment]
@rnveach , @nrmancuso , @pbludov , @strkkk , please share your thoughts

**Comment by rnveach** [comment]
Why not switch it to Javadoc's method of handling parsing failures and instead print a violation?

**Comment by romani** [comment]
It still very complicated concept to suppress. Most users do not know that we have filters and how to manage them.
As they switched to new jdk, and what to proceed with jdk, better to not force them to update suppression each time someone what to use new syntax.

**Comment by nrmancuso** [comment]
From the description, it seems to me that this exact behavior can be achieved by the `haltOnException` property and a `SuppressionSingleFilter` modlue:

```
➜  src cat config.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
        "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
  <module name="TreeWalker">
    <module name="AbbreviationAsWordInName"/>
  </module>
</module>

➜  src cat Test.java 
public clazz Test {}

➜  src java -jar checkstyle-10.4-all.jar -c config.xml Test.java
Starting audit...
com.puppycrawl.tools.checkstyle.api.CheckstyleException: Exception was thrown while processing Test.java
        at com.puppycrawl.tools.checkstyle.Checker.processFiles(Checker.java:307)
        at com.puppycrawl.tools.checkstyle.Checker.process(Checker.java:224)
        at com.puppycrawl.tools.checkstyle.Main.runCheckstyle(Main.java:415)
        at com.puppycrawl.tools.checkstyle.Main.runCli(Main.java:338)
        at com.puppycrawl.tools.checkstyle.Main.execute(Main.java:195)
        at com.puppycrawl.tools.checkstyle.Main.main(Main.java:130)
Caused by: com.puppycrawl.tools.checkstyle.api.CheckstyleException: IllegalStateException occurred while parsing file /home/nick/IdeaProjects/tester/src/Test.java.
        at com.puppycrawl.tools.checkstyle.JavaParser.parse(JavaParser.java:105)
...
org.antlr.v4.runtime.atn.ParserATNSimulator.adaptivePredict(ParserATNSimulator.java:396)
        at com.puppycrawl.tools.checkstyle.grammar.java.JavaLanguageParser.typeDeclaration(JavaLanguageParser.java:662)
        ... 11 more
Checkstyle ends with 1 errors.

```

```
➜  src cat config.xml                                           
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
        "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
  <property name="haltOnException" value="false"/>
  <module name="SuppressionSingleFilter">
    <property name="message" value=".*IllegalStateException occurred while parsing file.*"/>
  </module>
  <module name="TreeWalker">
    <module name="AbbreviationAsWordInName"/>
  </module>
</module>

➜  src cat Test.java                                            
public clazz Test {}

➜  src java -jar checkstyle-10.4-all.jar -c config.xml Test.java
Starting audit...
Audit done.


```

I think that this is fine, since a user finding either property (haltOnException or new property) will fall into one of two cases:

1. User reports error on new syntax, and we point them to this approach, or
2. User comes to report syntax, sees an issue already exists, and sees this config shared by a maintainer in existing issue.

We could certainly expand examples under Checker to include the above config and example, since this approach could also be extended to include other exceptions as well.

**Comment by romani** [comment]
new property will swallow only parse exception, like no exception at all.
existing property continue on all exceptions.

**Comment by romani** [comment]
so workaround is
```
<module name="Checker">
  <property name="haltOnException" value="false"/>
  <module name="SuppressionSingleFilter">
    <property name="message" value=".*IllegalStateException occurred while parsing file.*"/>
  </module>
....
```

Lets ask users to test it on real code.

**Comment by nrmancuso** [comment]
Ok, we have a nuance with workaround at  https://github.com/checkstyle/checkstyle/issues/12542#issuecomment-1362261537: https://github.com/checkstyle/checkstyle/issues/12507#issuecomment-1366683265

If we make new property, we can overcome this limitation, and should include testing of module order in config.

**Comment by romani** [comment]
and we will make effect of exception to be scoped to Treewalker.

We have options to make (same a in Checker, but we did this option mostly for our testing tool, not as user request)
```
<module name="TreeWalker">
  <property name="haltOnException" value="false"/>
```
or smth like:
```
<module name="TreeWalker">
  <property name="skipFileOnParseExcepion" value="false"/>
```

**Comment by rnveach** [comment]
Is there any reason we can't synch behavior between javadoc and java in this regards?

**Comment by romani** [comment]
I see no problems to have similar/additional property for javadoc parser.

I do not think we should mix them, as javadoc parsing problem is too frequent.

**Comment by romani** [comment]
@rnveach , @nrmancuso , @pbludov , @strkkk  please share your opinion/vote on this issue.

**Comment by nrmancuso** [comment]
Ok, let me try to sum up our discussion:

1. We will introduce a new property called `skipFileOnJavaParseException`:
```
<module name="TreeWalker">
  <property name="skipFileOnJavaParseException" value="false"/>

```

2. We will introduce a new property called `skipFileOnJavadocParseException`:
```
<module name="TreeWalker">
  <property name="skipFileOnJavadocParseException" value="false"/>

```

3. Update documentation accordingly

I agree with all above points, due to nondeterminism of module execution blocking solution at https://github.com/checkstyle/checkstyle/issues/12507#issuecomment-1366683265.

I am good to introduce new properties, however I feel from a user standpoint that I would rather create a `BeforeExecutionExclusionFileFilter` for each offending file. With the new properties, you really don't know what files are being skipped.

**Comment by romani** [comment]
> We will introduce a new property

Yes.

BeforeExecutionExclusionFileFilter is not very convenient in real project you have to love checkstyle to always update its config.
Most team members  don't care and want to spend less time on configuring checkstyle. Especially with new syntax, as all team members try to use new feature to feel how it works.

Checkstyle is tool on a side, it is not main tool. It should not block users to do what they want. If user ok to skip validation on whole file (many files) in favor of new jdk syntax, checkstyle should not block people.

I would do each properties in separate issues, I still have doubts that users need javadoc, as nobody complained on this. 

For logging, we do not do logging, I am not sure how it will work in ant/maven/gradle , we can improve on logging separately.

**Comment by cowwoc** [comment]
Any update on this issue?

**Comment by romani** [comment]
@cowwoc , please try https://github.com/checkstyle/checkstyle/issues/12542#issuecomment-1365367872

@rnveach , please approve issue if you agree. Looks like I and @nrmancuso is in agreement.

We can skip `name="skipFileOnJavadocParseException` now and create separate issue on it to make it to happen sooner.

**Comment by cowwoc** [comment]
@romani JDK 20, and yes... this is the issue: https://github.com/checkstyle/checkstyle/issues/13279

**Comment by romani** [comment]
@cowwoc , Does workaround work for you ?

**Comment by rnveach** [comment]
@romani Let me know when first post contains all final details I am approving.

**Comment by romani** [comment]
@rnveach , description is already in final state.

> skipFileOnJavadocParseException

better to not make it now, as we create parser in each javadoc Check, so it is not treewalker and it is more complicated. We will think on this some time later on, if demand on this grow.

**Comment by cowwoc** [comment]
@romani I believe it did.

**Comment by aayushRedHat** [comment]
I will take this

**Comment by Lmh-java** [comment]
I am on this

**Comment by nrmancuso** [comment]
This is a big win for users, thanks to @mahfouz72 for getting this across the finish line! 🚀
